src/librustc_resolve/lib.rs

```diff
@@ -3273,6 +3274,38 @@ impl<'a> Resolver<'a> {
                     let traits = self.get_traits_containing_item(item_name, ns);
                     self.trait_map.insert(id, traits);
                 }
+
+                // check if we wrote `str::from_utf8` instead of `std::str::from_utf8`
+                let item_span = path.iter().last().map(|segment| segment.ident.span)
+                    .unwrap_or(span);
+                let mut std_path = vec![Segment::from_ident(Ident::from_str("std"))];
+                std_path.extend(path);
+                match self.resolve_qpath_anywhere(
+                    id,
+                    None,
+                    &std_path,
+                    ns,
+                    span,
+                    source.defer_to_typeck(),
+                    source.global_by_default(),
+                    crate_lint,
+                    true,
+                ) {
+                    Some(resolution) if resolution.unresolved_segments() == 0 => {
+                        debug!("accessed item from `std` submodule as a bare type {:?}", std_path);
+                        let mut hm = self.session.confused_type_with_std_module.borrow_mut();
+                        hm.insert(
+                            item_span,
+                            span,
+                        );
```

Comment:
Nit: I don't think even rustfmt would format this on multiple lines (and rustfmt is known to be vertical space hungry).